### PR TITLE
Capture section and chapter in the metadata for epubs under `epub_section`

### LINKED ISSUE #459: Capture section and chapter in the metadata for epubs
The goal of this issue is to capture section and chapter in the metadata for epubs. Currently we're only capturing filename and page number.

### PR DESCRIPTION
**Pull request by Coniferish**
## Summary
Closes #459
Capture section and chapter in the metadata for epubs under `epub_section`
Adds associated tests

## Testing
```
from unstructured.partition.epub import partition_epub

filename="example-docs/winter-sports.epub"
elements = partition_epub(filename)
elements[0].metadata
```

### CODE REVIEW

File: unstructured/partition/epub.py
Comment (MthwRobinson) [inline unstructured/partition/epub.py:77]: We shouldn't have a bare `except` here. Are you hitting an error that you need to handle when testing this?

File: unstructured/partition/html.py
Comment (MthwRobinson) [inline unstructured/partition/html.py:74]: Why did this block and the later block for last modification date get added?

File: unstructured/staging/base.py
Comment (MthwRobinson) [inline unstructured/staging/base.py:164]: There are a couple of files with pure formatting changes like this that shouldn't be included in this PR.

File: unstructured/documents/elements.py
Comment (MthwRobinson) [inline unstructured/documents/elements.py:147]: Let's make this `section` instead of `epub_section` since other document types may have section information as well.

File: unstructured/file_utils/filetype.py
Comment (MthwRobinson) [inline unstructured/file_utils/filetype.py:506]: Why do we need to pass down `**kwargs` here now?

File: unstructured/ingest/connector/google_drive.py
Comment (MthwRobinson) [inline unstructured/ingest/connector/google_drive.py:79]: Please revert the change here and in `onedrive.py` since that's not related to the EPUB work.

**Review by MthwRobinson — CHANGES_REQUESTED** [review]
Added a couple of inline comments to the draft

File: unstructured/staging/base.py
Comment (Coniferish) [reply unstructured/staging/base.py:164]: I added these because checks were failing. Should I move these to a different PR and have that merged before merging this one?

File: unstructured/partition/epub.py
Comment (Coniferish) [reply unstructured/partition/epub.py:77]: I don't remember off the top of my head, but I don't believe it was raising an error. I'll remove it.

File: unstructured/partition/html.py
Comment (Coniferish) [reply unstructured/partition/html.py:74]: With my changes, `partition_epub` doesn't make a call to `convert_and_partition_html` any more and instead directly calls `partition_html` and passes the epub sections to it as text. Because of that, I added calls to `get_last_modified_date_from_file` and `get_last_modified_date` in `partition_epub` and pass those on to `html_partition`. Since the epub info is being pass to `partition_html` as text and `last_modification_date`, is being passed as an argument, I added that logic to get it from the `kwargs`

File: unstructured/file_utils/filetype.py
Comment (Coniferish) [reply unstructured/file_utils/filetype.py:506]: Because the `**kwargs` gets passed to `_add_element_metadata`, which is where the `epub_section` kwarg that's been passed down is added to the metadata

File: unstructured/staging/base.py
Comment (MthwRobinson) [reply unstructured/staging/base.py:164]: Pulling in the changes from #986 (once that's merged) will take care of this.

File: unstructured/partition/html.py
Comment (MthwRobinson) [reply unstructured/partition/html.py:74]: Instead of modifying that here, you'll want logic similar to the following to pull that information from the epub file in `partition_epub`.

https://github.com/Unstructured-IO/unstructured/blob/331c7faf38984d0ce29d920f6ac51d5071f6c0c5/unstructured/partition/html.py#L147-L164

File: unstructured/partition/html.py
Comment (MthwRobinson) [inline unstructured/partition/html.py:119]: This should stay as it was, #997 changed `metadata_date` to `metadata_last_modified`.

File: unstructured/partition/epub.py
Comment (MthwRobinson) [inline unstructured/partition/epub.py:72]: `metadata_date` you can get rid of now

File: unstructured/partition/html.py
Comment (Coniferish) [reply unstructured/partition/html.py:119]: Yah, I'm going through and fixing that with the rebase--that happened because of a recent commit

File: unstructured/partition/epub.py
Comment (MthwRobinson) [inline unstructured/partition/epub.py:63]: Do you know if the the TOC items corresponds to? Is it a chapter or something or does it depend on the book?

Comment (Coniferish) [reply unstructured/partition/epub.py:63]: I'm not 100% clear on that to be honest, but I know it includes more than just the "contents" section of a book. For example, the `winter-sports.epub` book we use for testing has the following toc items (plus the rest of the chapters):
'The Project Gutenberg eBook of Winter Sports in Switzerland, by E. F. Benson', 'WINTER SPORTS IN SWITZERLAND', 'CONTENTS', 'ILLUSTRATIONS', 'CHAPTER I THE SUN-SEEKER', etc.

And the final element in the toc items does include the text up to the end of the book (at least in this case--unsure how much variability there would be between different epub books)

Comment (MthwRobinson) [reply unstructured/partition/epub.py:63]: Gotcha, may be worth trying a couple of epubs from Project Gutenberg just to see what a few of them look like. What do you think about also added a section number to go along with the section name? Since there's no guarantee section names will be unique.

- https://www.gutenberg.org/

Comment (Coniferish) [reply unstructured/partition/epub.py:63]: Would we want to make partitioning via toc default and then have it partition via the existing way be an option users could pick via an argument?

Comment (MthwRobinson) [reply unstructured/partition/epub.py:63]: I don't think so. This way seems better since it doesn't require a system dependency (`pandoc`) and doesn't require the overhead of converting the doc. They can alway use `convert_and_partition_html` directly if they do want to convert with `pandoc` first and then partition.

Comment (Coniferish) [reply unstructured/partition/epub.py:63]: for this [book](https://www.gutenberg.org/cache/epub/6167/pg6167.html), I have the following toc items (and have verified that we're partitioning the beginning and ending of it):
(Pdb) titles
['CONTENTS', 'CHAPTER I COMMON SENSE FARMING 1. Wealth, Land and Labor. 2. Money. 3. Sources of Wealth. 4. The Farmer, a Producer, and Seller. 5. Business Methods Essential.', 'INTRODUCTION', 'CHAPTER I', 'CHAPTER II', 'CHAPTER III', 'CHAPTER IV', 'CHAPTER V', 'CHAPTER VI', 'CHAPTER VII', 'CHAPTER VIII', 'CHAPTER IX', 'CHAPTER X', 'CHAPTER XI', 'CHAPTER XII', 'CHAPTER XIII', 'CHAPTER XIV', 'CHAPTER XV', 'CHAPTER XVI', 'CHAPTER XVII', 'CHAPTER XVIII', 'CHAPTER XIX', 'CHAPTER XX', 'CHAPTER XXI', 'CHAPTER XXII', 'CHAPTER XXIII', 'CHAPTER XXIV', 'CHAPTER XXV', 'CHAPTER XXVI', 'CHAPTER XXVII', 'CHAPTER XXVIII', 'THE FULL PROJECT GUTENBERG LICENSE']
(Pdb) tuple_titles
['BUSINESS HINTS FOR MEN AND WOMEN']

and this [book](https://www.gutenberg.org/cache/epub/12511/pg12511-images.html)
(Pdb) titles
["The Project Gutenberg eBook of Blackwood's Edinburgh Magazine — Volume 53, No. 332, June, 1843", "BLACKWOOD'S EDINBURGH MAGAZINE", 'NO. CCCXXXII. JUNE, 1843. VOL. LIII.', 'CONTENTS.', "A PASSAGE IN THE LIFE OF A MAÎTRE-D'ARMES", "MR BAILEY'S REPLY TO AN ARTICLE IN BLACKWOOD'S MAGAZINE.", "THE ATTORNEY'S CLERK IN THE MONK'S HOOD.", 'COMMERCIAL POLICY—RUSSIA.', 'FOOTNOTES.', 'INDEX TO VOL. LIII.', 'END OF VOL. LIII.', 'THE FULL PROJECT GUTENBERG LICENSE']
(Pdb) tuple_titles
['MARSTON; OR, THE MEMOIRS OF A STATESMAN.', 'THE VIGIL OF VENUS.', 'CHAPTERS OF TURKISH HISTORY. RISE OF THE KIUPRILI FAMILY—SIEGE OF CANDIA.', 'AMMALÁT BEK.', 'THE VICAR OF WAKEFIELD.', 'IGNACIO GUERRA AND EL SANGRADOR;', "MEMORANDUMS OF A MONTH'S TOUR IN SICILY."]

Comment (Coniferish) [reply unstructured/partition/epub.py:63]: I am just finding, though, that the `toc` items are overlapping in some circumstances--the `html_content` for the second book I listed is the same for the first couple of items in the `toc`...

Comment (MthwRobinson) [reply unstructured/partition/epub.py:63]: Roger that - thanks!

**Comment by MthwRobinson** [comment]
@Coniferish - You'll need to add the epubs requirements here to fix those test failures in CI.

https://github.com/Unstructured-IO/unstructured/blob/a23d75a292ef953afb6d802595e8f717b210e615/Makefile#L92-L93

**Comment by MthwRobinson** [comment]
And also add an `install-epub` make target, like this one for docx

https://github.com/Unstructured-IO/unstructured/blob/a23d75a292ef953afb6d802595e8f717b210e615/Makefile#L60-L62

**Comment by Coniferish** [comment]
> @Coniferish - You'll need to add the epubs requirements here to fix those test failures in CI.
> 
> https://github.com/Unstructured-IO/unstructured/blob/a23d75a292ef953afb6d802595e8f717b210e615/Makefile#L92-L93

I noticed that `install-docx` is listed twice in this line. While I'm adding `install-ebooklib`, should I remove the duplicate?

File: Makefile
Comment (MthwRobinson) [inline Makefile:64]: Let's change this to `install-epub`

File: test_unstructured/partition/test_epub.py
Comment (MthwRobinson) [inline test_unstructured/partition/test_epub.py:14]: Can we assert what the name of the section should be instead of just check that it's not `None`? One other good test here would be to get `all_sections = set([element.metadata.section for element in elements])` and check that all the sections we're expecting are present.

File: unstructured/documents/elements.py
Comment (MthwRobinson) [inline unstructured/documents/elements.py:146]: Can we describe this as `# Document section fields` since other document types could have sections as well?

File: unstructured/partition/epub.py
Comment (MthwRobinson) [inline unstructured/partition/epub.py:65]: Instead of constructing `html_items` in a list comprehension, can we do:

```python
for item in book.items:
    if not isinstance(item, epub.EpubHtml):
        continue
    <rest of stuff>
```

Also, what else is in `book.items` aside from `EpubHtml`? Is that anything we want to capture?

File: unstructured/partition/epub.py
Comment (MthwRobinson) [inline unstructured/partition/epub.py:69]: We should accept an `encoding` kwarg in `partition_epub` and decode using the appropriate encoding. @christinestraub has done a lot of work on encoding already and may be able to help with that.

https://github.com/Unstructured-IO/unstructured/blob/main/unstructured/file_utils/encoding.py

File: test_unstructured/partition/test_epub.py
Comment (Coniferish) [reply test_unstructured/partition/test_epub.py:14]: That's what the test was before, but not all elements have a section field--the issue I was running into before was using the toc as the primary means of parsing, but it's not totally reliable in that multiple toc elements point to the same text and some sections of text aren't included in the toc. I defaulted to having those parts that don't have a corresponding toc element to being `None`, but that could be changed to "NA" or something else.

That said, the toc isn't totally reliable it seems--it doesn't fully align with the actual chapters in the book (which is why a chunk of chapter 2 of the test epub book was missing in the elements when I was just using the toc for parsing)

File: unstructured/partition/epub.py
Comment (Coniferish) [reply unstructured/partition/epub.py:65]: There are `EpubLink`, `EpubImage`, `EpubNcx` (epub page navigation info) and `EpubItem` (which seems to just be fomatting/css info). Do we want to capture the images?

File: unstructured/partition/epub.py
Comment (Coniferish) [reply unstructured/partition/epub.py:69]: I believe I took this snippet `item.get_content().decode()` from Camgreenburg's function, but looking at it more closely, it doesn't seem the `.decode()` part is actually needed (or used--I stepped through `pdb` and didn't see it called and can't find a definition for it through VSC). 
I did find the `.get_content()` definition in ebiooklib, but it doesn't accept an argument for encoding:
```    
def get_content(self, default=None):
        """
        Returns content for this document as HTML string. Content will be of type 'str' (Python 2)
        or 'bytes' (Python 3).

        :Args:
          - default: Default value for the content if it is not defined.

        :Returns:
          Returns content of this document.
        """
```

Would you want to create a fork so we can update this to accept an encoding argument?

Comment (Coniferish) [reply unstructured/partition/epub.py:69]: Nope, I was mistaken about the `.decode()`. It is being used--I must have missed it when stepping through `pdb` (though I'm not sure why it's not being mapped to a definition in VSC)

File: test_unstructured/partition/test_epub.py
Comment (MthwRobinson) [reply test_unstructured/partition/test_epub.py:14]: Just looking for something along the lines of this (using the actual expected values for the test document)

```python
assert elements[0].metadata.section == "ch1"

all_sections = set([el.metadata.section for el in elements])
assert all_sections = expected_sections
```

**Review by MthwRobinson — DISMISSED** [review]
Added a few more inline comments

File: unstructured/partition/html.py
Comment (Coniferish) [reply unstructured/partition/html.py:143]: I don't think I made that change... code blame is showing cragwolfe made that change last week. I may be mistaken, though. I had to resolve some merge conflicts when I rebased onto master. 
https://github.com/Unstructured-IO/unstructured/pull/997/files#diff-00fb69efff2377f4bdcfbe3fceb67211ea891f72285dc4c502609865c37bb3aa

File: unstructured/partition/epub.py
Comment (Coniferish) [reply unstructured/partition/epub.py:69]: When I use `detect_file_encoding(filename)` on the test epub, I'm getting 'iso-8859-1'. I am wondering if it's necessary for me to use that, though, since it seems to slow down the test significantly. Could it be hard-coded?

This is what I have added locally for encoding, but was hoping to meet with or hear from Christine to take a look at it:
```
        if encoding:
            item_content = item.get_content().decode(encoding)
        else:
            item_content = item.get_content().decode(formatted_encoding)
```

File: unstructured/partition/epub.py
Comment (Coniferish) [reply unstructured/partition/epub.py:43]: I thought a test was failing without it, but I just checked and everything is passing. Removed it

File: unstructured/partition/epub.py
Comment (christinestraub) [reply unstructured/partition/epub.py]: @MthwRobinson @Coniferish I noticed that the `EpubHtml.get_content` method always returns `utf-8` encoded content (type `bytes`). https://github.com/aerkalov/ebooklib/blob/master/ebooklib/epub.py#L437
```
tree_str = etree.tostring(tree, pretty_print=True, encoding='utf-8', xml_declaration=True)
```
IMO, we don't have to use the `detect_file_encoding` as long as we use the `EpubHtml.get_content`.

Comment (Coniferish) [reply unstructured/partition/epub.py]: Sounds good. I suspected that, but am glad the second opinion!

Comment (MthwRobinson) [reply unstructured/partition/epub.py]: Sounds good - thanks for checking that out!

File: unstructured/partition/html.py
Comment (MthwRobinson) [inline unstructured/partition/html.py:126]: In this file it looks like the only actual change is adding `**kwargs`. Could you clean up the rest of the diffs in here?

File: test_unstructured/partition/test_epub.py
Comment (MthwRobinson) [inline test_unstructured/partition/test_epub.py:8]: Are these actually out of order in the output or is this just because it's a set? Could you order these regardless so they're easier to read?

Comment (MthwRobinson) [reply test_unstructured/partition/test_epub.py:8]: Looks like the order is correct when I run the sample code, so just need to reorder these for readability.

Comment (Coniferish) [reply test_unstructured/partition/test_epub.py:8]: Sorry about that. It's just what I copied when I ran it as a set

File: unstructured/partition/html.py
Comment (Coniferish) [reply unstructured/partition/html.py:126]: Looking into this now

File: unstructured/partition/html.py
Comment (MthwRobinson) [inline unstructured/partition/html.py:156]: I think we still want `last_modification_date` initialized here

**Comment by MthwRobinson** [comment]
Looking good, just pull in the latest from `main` and re-run `pip-compile` to clean up the conflicts. And add one small comment. Otherwise looks good to merge once the conflicts are cleaned up.

**Comment by Coniferish** [comment]
> Looking good, just pull in the latest from `main` and re-run `pip-compile` to clean up the conflicts. And add one small comment. Otherwise looks good to merge once the conflicts are cleaned up.

@MthwRobinson I'm resolving conflicts in the requirements and wondering what to do in this case where I don't have these requirements locally, but they're in main. Is it safe to remove them? I've run `make check-deps` and everything seems to be in order on my end.
<<<<<<< 459_fixed
=======
importlib-metadata==6.8.0
    # via
    #   jupyter-client
    #   jupyter-lsp
    #   jupyterlab
    #   jupyterlab-server
    #   nbconvert
importlib-resources==6.0.1
    # via
    #   jsonschema
    #   jsonschema-specifications
    #   jupyterlab
    #   notebook
>>>>>>> main

**Review by cragwolfe — APPROVED** [review]

**Pull request merged**
